presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java

```diff
@@ -357,57 +344,30 @@ public void moreMemoryAvailable()
         memoryFuture.get().set(null);
     }
 
-    public void transferMemoryToTaskContext(long taskBytes)
-    {
-        long bytes = memoryReservation.getAndSet(0);
-        driverContext.transferMemoryToTaskContext(bytes);
-
-        TaskContext taskContext = driverContext.getPipelineContext().getTaskContext();
-        if (taskBytes > bytes) {
-            try {
-                taskContext.reserveMemory(taskBytes - bytes);
-            }
-            catch (ExceededMemoryLimitException e) {
-                taskContext.freeMemory(bytes);
-                throw e;
-            }
-        }
-        else {
-            taskContext.freeMemory(bytes - taskBytes);
-        }
-    }
-
-    public void setMemoryReservation(long newMemoryReservation)
-    {
-        checkArgument(newMemoryReservation >= 0, "newMemoryReservation is negative");
-
-        long delta = newMemoryReservation - memoryReservation.get();
-
-        if (delta > 0) {
-            reserveMemory(delta);
-        }
-        else {
-            freeMemory(-delta);
-        }
-    }
-
-    public boolean trySetMemoryReservation(long newMemoryReservation)
+    /**
+     * This method is <b>not</b> thread safe. If multiple threads work on the same
+     * <code>transferredBytesMemoryContext</code> instance it's possible that one of the writes
+     * is lost by the last <code>transferredBytesMemoryContext.setBytes()</code> call.
```

Comment:
It could be averted by synchronizing on `target` for the whole duration of `com.facebook.presto.memory.context.SimpleLocalMemoryContext#transferMemory`